pkg/rhcos/builds.go

```diff
@@ -36,42 +35,77 @@ type metadata struct {
 	OSTreeVersion string `json:"ostree-version"`
 }
 
-func fetchLatestMetadata(ctx context.Context, channel string) (metadata, error) {
-	build := buildName
-	var err error
-	if build == "" {
-		build, err = fetchLatestBuild(ctx, channel)
-		if err != nil {
-			return metadata{}, errors.Wrap(err, "failed to fetch latest build")
+// rhcosBuildID is a version number e.g. 410.8.20180312.0
+var rhcosBuildID string
+
+// rhcosBuildChannel is a name for a stream of builds
+var rhcosBuildChannel string
+
+func fetchRHCOSBuild(ctx context.Context) (metadata, error) {
+	// Example uses:
+	// env OPENSHIFT_INSTALL_RHCOS_OVERRIDE=latest
+	// env OPENSHIFT_INSTALL_RHCOS_OVERRIDE=410.8.20190314.0
+	// env OPENSHIFT_INSTALL_RHCOS_OVERRIDE=testing/walters@410.8.20190312.0
+	if rc, ok := os.LookupEnv("OPENSHIFT_INSTALL_RHCOS_OVERRIDE"); ok && rc != "" {
+		logrus.Warn("Found override for RHEL CoreOS. Please be warned, this is not advised")
+		parts := strings.Split(rc, "@")
+		if len(parts) == 2 {
+			rhcosBuildChannel = parts[0]
+			rhcosBuildID = parts[1]
+		} else {
+			rhcosBuildChannel = defaultChannel
+			rhcosBuildID = parts[0]
 		}
 	}
+	var body []byte
+	// If there's no override, load our hardcoded version
+	if rhcosBuildID == "" {
+		// TODO(walters) for now
+		rhcosBuildChannel = "releases/maipo"
+		file, err := data.Assets.Open("rhcos.json")
+		if err != nil {
+			return metadata{}, err
+		}
+		defer file.Close()
+		body, err = ioutil.ReadAll(file)
+	} else {
+		var build string
+		var err error
+		if rhcosBuildID == "latest" {
+			build, err = fetchLatestBuild(ctx, rhcosBuildChannel)
+			if err != nil {
+				return metadata{}, errors.Wrap(err, "failed to fetch latest build")
+			}
+		} else {
+			build = rhcosBuildID
+		}
+		url := fmt.Sprintf("%s/%s/%s/meta.json", baseURL, rhcosBuildChannel, build)
```

Comment:
I'd rather just stick with our existing `OPENSHIFT_INSTALL_OS_IMAGE_OVERRIDE` for overrides.  Coupled with the bundled metadata for the pinned RHCOS that this PR is adding, that would mean we could drop this JSON request entirely.  Different override approaches have been floated before in #1168 and #1402 though, so there's clearly some workflow where "supply the override AMI (or QCOW2 image URI, etc.)" is not seen as sufficient.  Can you point me at that place, so I can float an approach that uses `OPENSHIFT_INSTALL_OS_IMAGE_OVERRIDE`?

Reply:
I think anyone who wants to test something related to RHCOS will find it *far* easier to supply a version number rather than parsing the version number and getting the AMI or extracting the qcow2 manually.  I can very much say this is the case for me.

Further, I expect us to at some point have CI that overrides this.  "Test this release payload with updated rhcos bootimage" is a valid thing to want to do for the same reason we want to override release payloads.

Reply:
Really we need both.  The existing image override is useful to test an *unofficial* bootimage.

In fact just now I used both - tested with the latest rhcos from our buildsystem, hit an issue, did a local build with a test fix, then used the local override to point to `file:///`.

Reply:
> I think anyone who wants to test something related to RHCOS will find it far easier to supply a version number rather than parsing the version number and getting the AMI or extracting the qcow2 manually.

How many people need this, though?

> Further, I expect us to at some point have CI that overrides this. "Test this release payload with updated rhcos bootimage" is a valid thing to want to do for the same reason we want to override release payloads.

Sure, but my preferred approach for this *is* "override the release image" (via #1286).

> Really we need both. The existing image override is useful to test an *unofficial* bootimage.

It can be used for testing official boot images too, right?  The only issue is whether this lookup needs to be built into the installer, or if you can use external tool like [this script][1].

[1]: https://github.com/openshift/installer/pull/1168#issuecomment-459501811

Reply:
> Sure, but my preferred approach for this is "override the release image" (via #1286).

You know I agree :100: % with that - particularly because if e.g. say one is overriding the image thinking to test a kernel change to e.g. a NIC driver for a new cloud platform, it will be a surprise probably for most people that that change gets overwritten quickly by `machine-os-content` - which you need a custom release payload to change.  (And we need to make changing both ergonomic, but that's another issue)

But my real argument here is:

> The only issue is whether this lookup needs to be built into the installer

The code's *already there* though!  It's just missing a few config knobs.

> How many people need this, though?

It's not just the number of people - it's also that I don't want lots of places parsing that release schema - while we're unlikely to change it soon, again the installer already has the code, so let's just make it more useful.

Reply:
> The code's already there though! It's just missing a few config knobs.

But this is our chance to get rid of it, and "more knobs" feels like doubling down on a known dead-end approach :p.

Reply:
> But this is our chance to get rid of it, and "more knobs" feels like doubling down on a known dead-end approach :p.

I don't think we're going to deploy the dependencies necessary in #1286 anytime soon.

Also, this code would be useful for us to deal with the Ignition 3 change.